### ISSUE TITLE
IES Lights using the light projector system

### ISSUE THREAD

**Issue by fire** [godotengine/godot-proposals#715]
**Describe the project you are working on:**
Not related to my project.

**Describe the problem or limitation you are having in your project:**
Lights in Godot Engine are clunky and look the same.

There are existing IES profiles that give varied and exciting light types.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
It would make artwork and level design improved if IES light profiles can be imported directly.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
1. Read IES profile
2. Generate a format the light projector feature can use
3. Import the format as a light projector

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

No, the convenience of dragging a profile to be a light project is the feature.

**Is there a reason why this should be core and not an add-on in the asset library?:**
In theory this can be a gdnative plugin, but ies profiles is a important visual tool in lighting a scene.

**Comment by Calinou** [comment]
Out of curiosity, are there freely-licensed IES profiles out there? Not just freely downloadable, but under a public domain-like license or a free Creative Commons license. This would be useful for a a demo project.

**Comment by fire** [comment]
https://renderman.pixar.com/ies-profiles I found this one.

**Comment by fire** [comment]
https://seblagarde.wordpress.com/tag/ies-specification/ Described how Frostbite implemented IES lights.

**Comment by clayjohn** [comment]
Also Filament has a great explanation https://google.github.io/filament/Filament.md.html#lighting/directlighting/photometriclights

**Comment by orbatos** [comment]
At [ies Library](https://ieslibrary.com) (https://ieslibrary.com), which started early this month, they currently have more than 100k IES files and tools to convert to LDT. They are also planning a Blender plugin to access the library, a type of utility that could be valuable for Godot as well.

**Comment by Migui2611** [comment]
I find this very necessary. It would be a great leap forward in Godot rendering capabilities. IES lights are a important step for achieving realism, and would put Godot in the ArchViz scope.

**Comment by fire** [comment]
See also https://github.com/KhronosGroup/glTF/pull/2200

**Comment by Calinou** [comment]
Regarding the IES rendering itself, isn't a dedicated 1D **Attenuation Texture** property needed? The projector texture is applied in 2D space, but it doesn't apply to attenuation light itself (when light goes further away from the origin). Projectors also require shadows to be enabled, which can be too demanding for some game use cases (but I assume you always have shadows enabled for archviz).

**Comment by fire** [comment]
Open the ies light and assign as a panorma.

**Comment by Nikoraito** [comment]
I would REALLY love to see this

**Comment by Calinou** [comment]
@Nikoraito Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by fire** [comment]
> # 4.5 Photometric Lights Summary
> 
> Photometric lights utilize a photometric profile to describe their intensity distribution, which is stored in a photometric file. There are two common formats: IES (.ies) and EULUMDAT (.ldt).
> 
> - **IES**: Created by the Illuminating Engineering Society for electronic transfer of photometric data over the web. It's widely used in North America and Europe.
> - **EULUMDAT**: The European standard and industry standard photometric data file in Europe.
> 
> Many lighting manufacturers provide these files freely on their websites. Both formats store luminous intensity for various angles, measured using light sensors spread spherically around a light source.
> 
> The spherical coordinate system used to describe light distribution is referred to as "the photometric web". There are three types of photometric webs:
> 
> 1. Type A: For automotive headlamp and signal lights
> 2. Type B: For adjustable outdoor area and sports lighting luminaires
> 3. Type C: For architectural and road lights (most commonly used in computer graphics)
> 
> The IES format stores luminous intensity values in candela, while the EULUMDAT format stores luminous intensity values in candela per total kilo-lumens emitted by the light.
> 
> In Frostbite, only the IES format is supported due to its widespread use in computer graphics. A photometric profile, created from IES files, can be directly applied on a point or a spot light. The IES profile can be used for describing the light intensity and can be adjusted with a multiplier.
> 
> When creating a new light profile, the spherical photometric function is reconstructed and sampled to fill a 2D texture with a spherical parametrization. In shaders, the 2D texture is evaluated and applied as an attenuation.
> 
> ```csharp
> float getIESProfileAttenuation(float3 L, ShadowLightInfo light)
> {
>     // Sample direction into light space
>     float3 iesSampleDirection = mul(light.worldToLight, -L);
> 
>     // Cartesian to spherical
>     // Texture encoded with cos(phi), scale from -1 ->1 to 0 ->1
>     float phiCoord = (iesSampleDirection.z * 0.5f) + 0.5f;
>     float theta = atan2(iesSampleDirection.y, iesSampleDirection.x);
>     float thetaCoord = theta * FB_INV_TWO_PI;
>     float3 texCoord = float3(thetaCoord, phiCoord, light.lightIndex);
>     float iesProfileScale = iesTexture.SampleLevel(sampler, texCoord, 0).r;
> 
>     return iesProfileScale;
> }
> ```
> 
> IES profiles are more useful for architectural design than in games. However, they can be used for simulating complex shadows, similar to cookie textures but with a different parametrization.

https://seblagarde.files.wordpress.com/2015/07/course_notes_moving_frostbite_to_pbr_v32.pdf

I am encountering linkrot here is the relevant text.

**Comment by Mohammad9760** [comment]
any news on this?
is anyone working to implement this feature? if yes please share the progress.

**Comment by Calinou** [comment]
> any news on this?
is anyone working to implement this feature? if yes please share the progress.

To my knowledge, nobody is currently working on implementing this feature.

**Comment by fire** [comment]
My last work was here https://github.com/V-Sekai/godot/tree/salvage/ies-lights

**Comment by unfa** [comment]
I wonder - in the absence of such a feature, could a Gradient 1D texture be used with a Light shader to at least approximate this effect without using texture proiection (which needs shadows, which is costly)?

**Comment by Calinou** [comment]
> I wonder - in the absence of such a feature, could a Gradient 1D texture be used with a Light shader to at least approximate this effect without using texture proiection (which needs shadows, which is costly)?

Yes, a custom `light()` function in a spatial shader should be able to achieve this.

**Comment by MikeGO2611** [comment]
> I find this very necessary. It would be a great leap forward in Godot rendering capabilities. IES lights are a important step for achieving realism, and would put Godot in the ArchViz scope.

So, how things have changed, I've learn basic stuff about graphic pipeline and shaders, and as the one that commented this quote I have to correct myself. I no longer find this very necessary, in fact if implemented following the Godot philosophy it should be done in a plugin and not in the main engine.

Solutions provided here about using a GradientTexture2D or GradientTexture1D in the projection of a light are more optimal, the only issue is that there is no option to radially project the Gradient1D.

**Comment by MikeGO2611** [comment]
Using a Spotlight3D, a projector of a GradientTexture2D set to radial, and a GradientTexture1D, I'm able to simulate IES Profiles but with complete artistic control.

You can check the results atacched:

![GradientLight](https://github.com/user-attachments/assets/f759e6d9-c4a1-4124-8057-07f58dc4e3a7)
![GradientLightSettings](https://github.com/user-attachments/assets/c3c4cf91-aaab-4490-a95c-b73fb0182bcb)

**Comment by FlooferLand** [comment]
I really like that workaround, I think there still should be an official implementation though.
Some people might want to use IES profiles of real lights, the 1D gradient editor is kinda clunky, and that makes very symmetrical looking IES profiles.

Just throwing my hat in if anyone more skilled is able to implement this, but since IES files are just raw text files full of numbers, it would be nice if this hypothetical `IesTexture` resource allowed us to create our own IES files within Godot. That would have the side-benefit of allowing people to visually edit existing IES files imported from websites as well as using downloading Godot to use it as a best-in-class IES editing tool :)

_PS: You can save the resource from the workaround that was posted above as a tres file and the editor gives you pretty previews for it_

**Comment by atirut-w** [comment]
Can't one just render an IES profile to a cubemap projector?

**Comment by Calinou** [comment]
> Can't one just render an IES profile to a cubemap projector?

The projector is only half of the picture. The IES profile also defines light attenuation, which is relative to the distance from the light's origin. A projector texture cannot modify this in any way.

**Comment by the-simian** [comment]
> Out of curiosity, are there freely-licensed IES profiles out there? Not just freely downloadable, but under a public domain-like license or a libre Creative Commons license. This would be useful for a demo project.

Late to answer, but [CNDL](https://cndl.io/) An open source IES lighting creation tool, if that's useful here.

**Comment by unfa** [comment]
> Can't one just render an IES profile to a cubemap projector?
That'd be interesting.
Is it possible to project cubemaps with OmniLights?
Or reflection probes?

**Comment by atirut-w** [comment]
> Is it possible to project cubemaps with OmniLights?

The base Light3D class which OmniLight3D inherits have a property for projector textures, so in theory, it should work just fine. I've never tried it myself, though. If it works as I expected, one could write an IES importer that creates a texture.

**Comment by Calinou** [comment]
Rendering the IES to a cubemap will run into the same issue as https://github.com/godotengine/godot-proposals/issues/715#issuecomment-1288141406. It may not be a dealbreaker for some use cases, but it's worth thinking about.
